llama-index-integrations/indices/llama-index-indices-managed-postgresml/llama_index/indices/managed/postgresml/query.py

```diff
@@ -0,0 +1,193 @@
+from llama_index.core.prompts.base import ChatPromptTemplate, BasePromptTemplate
+from llama_index.core.prompts.mixin import PromptDictType, PromptMixinType
+from llama_index.core.schema import NodeWithScore, QueryBundle, TextNode
+from llama_index.indices.managed.postgresml.retriever import PostgresMLRetriever
+
+
+PROMPTS = {
+    "text_qa_template": ChatPromptTemplate(
+        message_templates=[
+            ChatMessage(content="You are a helpful chatbot", role=MessageRole.SYSTEM),
+            ChatMessage(
+                content="""Context information is below.
+---------------------
+{context_str}
+---------------------
+Given the context information and not prior knowledge,
+Query: {query_str}
+Answer: """,
+                role=MessageRole.USER,
+            ),
+        ]
+    )
+}
+
+
+class AsyncJsonGenerator:
+    def __init__(self, rag_stream_results) -> None:
+        self.rag_stream_results = rag_stream_results
+        self.rag_stream = None
+
+    def __iter__(self) -> "AsyncJsonGenerator":
+        return self
+
+    def __aiter__(self) -> "AsyncJsonGenerator":
+        return self
+
+    def __next__(self) -> str:
+        try:
+            return run_async_tasks([self.__anext__()])[0]
+        except StopAsyncIteration:
+            raise StopIteration
+
+    async def __anext__(self) -> str:
+        if not self.rag_stream:
+            self.rag_stream = self.rag_stream_results.stream()
+        result = await self.rag_stream.__anext__()
+        if len(result) > 0:
+            return result[0]
+        else:
+            return ""
+
+
+class PostgresMLQueryEngine(BaseQueryEngine):
+    """Retriever query engine for PostgresML."""
+
+    def __init__(
+        self,
+        retriever: PostgresMLRetriever,
+        streaming: Optional[bool] = False,
+        callback_manager: Optional[CallbackManager] = None,
+    ) -> None:
+        self._retriever = retriever
+        self._streaming = streaming
+        self._prompts = deepcopy(PROMPTS)
+        super().__init__(callback_manager=callback_manager)
+
+    @classmethod
+    def from_args(
+        cls,
+        retriever: PostgresMLRetriever,
+        **kwargs: Any,
+    ) -> "PostgresMLQueryEngine":
+        """Initialize a PostgresMLQueryEngine object.".
+
+        Args:
+            retriever (PostgresMLRetriever): A PostgresML retriever object.
+        """
+        return cls(retriever=retriever, **kwargs)
+
+    def with_retriever(self, retriever: PostgresMLRetriever) -> "PostgresMLQueryEngine":
+        return PostgresMLQueryEngine(
+            retriever=retriever,
+        )
+
+    def _query(self, query_bundle: QueryBundle, **kwargs) -> RESPONSE_TYPE:
+        """Answer a query."""
+        if self._streaming:
+            async_token_gen = run_async_tasks([self._do_query(query_bundle, **kwargs)])[
+                0
+            ]
+            return StreamingResponse(response_gen=async_token_gen.response_gen)
```

Comment:
slightly confused -- is the returned object here an async generator or a sync generator?

Reply:
It is actually both! It implements both __iter__ and __aiter__ and the other required methods.

Reply:
Oh neat!